### PR TITLE
Add fast quaternion implementation, minor changes to complex

### LINKED ISSUE #155: Implement quaternion algebra
**Reference implementations:**
https://github.com/amodeus-science/tensor/blob/ebeaaf9ec83ab6be2798a1548f131eeda9a37cd7/src/main/java/ch/ethz/idsc/tensor/qty/QuaternionImpl.java

https://github.com/apache/commons-numbers/blob/master/commons-numbers-quaternion/src/main/java/org/apache/commons/numbers/quaternion/Quaternion.java

### PR DESCRIPTION
**Pull request by CommanderTvis**
closes #155

### CODE REVIEW

**Comment by CommanderTvis** [comment]
This PR currently lacks tests, also complex . quaternion operations are needed

**Review by altavir — COMMENTED** [review]
Maybe move complex and quaternion to a separate module, say `kmath-complex`? It would help to maintain proper stability guarantees. I am not sure about that, but it does not make a lot of sense to keep them in the core.

**Comment by altavir** [comment]
Update plugin version to `0.7.3-1.4.30-RC` to fix tests.

**Comment by altavir** [comment]
Please resolve conflicts

File: examples/src/main/kotlin/kscience/kmath/operations/ComplexDemo.kt [hunk 1]
```diff
@@ -1,22 +1,24 @@
...
     // 2d element
-    val element = NDAlgebra.complex(2, 2).produce { (i,j) ->
+    val element = NDAlgebra.complex(2, 2).produce { (i, j) ->
         Complex(i.toDouble() - j.toDouble(), i.toDouble() + j.toDouble())
     }
     println(element)
 
     // 1d element operation
-    val result = with(NDAlgebra.complex(8)) {
+    val result = (NDAlgebra.complex(8)) {
```
Comment (altavir) [inline examples/src/main/kotlin/kscience/kmath/operations/ComplexDemo.kt:15]: Revert to more readable `with` please

File: kmath-complex/src/commonMain/kotlin/kscience/kmath/complex/Complex.kt [hunk 2]
```diff
@@ -42,74 +45,70 @@ private val PI_DIV_2 = Complex(PI / 2, 0)
 /**
  * A field of [Complex].
  */
-@OptIn(UnstableKMathAPI::class)
 public object ComplexField : ExtendedField<Complex>, Norm<Complex, Complex>, RingWithNumbers<Complex> {
-    override val zero: Complex = 0.0.toComplex()
-    override val one: Complex = 1.0.toComplex()
+    public override val zero: Complex by lazy { 0.0.toComplex() }
```
Comment (altavir) [inline kmath-complex/src/commonMain/kotlin/kscience/kmath/complex/Complex.kt:49]: Lazy will be much more expensive here than an assignment. I do not think that there is a difference, but probably better to keep it simple.

File: kmath-core/build.gradle.kts [hunk 3]
```diff
@@ -1,17 +1,27 @@
...
+kotlin.sourceSets {
+    commonMain {
+        dependencies {
+            api(project(":kmath-memory"))
+        }
+    }
+
+    commonTest {
+        dependencies {
+            api(project(":kmath-complex"))
```
Comment (altavir) [inline kmath-core/build.gradle.kts:17]: Move appropriate tests to kmath-complex and remove the dependency.

**Review by altavir — CHANGES_REQUESTED** [review]
Looks good. Only small changes are in order.

**Review by altavir — APPROVED** [review]

**Pull request merged**